File: 01.计算机科学/算法/leetcode/1.两数之和.java
/*
 * @lc app=leetcode.cn id=1 lang=java
 *
 * [1] 两数之和
 */

// @lc code=start
class Solution {
    public int[] twoSum(int[] nums, int target) {
        return func1(nums, target);
        // Map<Integer,Integer> map = new HashMap<Integer,Integer>();
        // for(int i = 0;i<nums.length;i++){
        //     if(map.containsKey(target - nums[i])!=null){
        //         return new int[2]{map.get(target-nums[i]),nums[i]};
        //     }else{
        //         map.Add(nums[i],i);
        //     }
        // }
        // return new int[0];
    }

    private int[] func1(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        int[] arr = new int[2];
        for (int i = 0; i < nums.length; i++) {
            int x = nums[i];
            if (map.get(x) != null) {// 直接找 目标差值 （target-x）
                arr[1] = i;
                arr[0] = map.get(nums[i]);
                return arr;
            } else {
                map.put((target - x), i);// 存索引
            }
        }
        return null;
    }
}
// @lc code=end

File: 01.计算机科学/算法/leetcode/1.两数之和.cs
/*
 * @lc app=leetcode.cn id=1 lang=csharp
 *
 * [1] 两数之和
 */

// @lc code=start
public class Solution
{
    //主要是利用Dic ,Key 存储 nums[i]  Value 存储 索引i
    public int[] TwoSum(int[] nums, int target)
    {
        Dictionary<int, int> dic = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            //! target - nums[i] 是比较巧妙的，用差来查找，Dic 是否已经存在 差。然后 Dic 的 value 存储的是 Index
            if (dic.ContainsKey(target - nums[i]) && dic[target - nums[i]] != i)
            {
                return new int[2] { dic[target - nums[i]], i };
            }
            if (dic.ContainsKey(nums[i]) == false)
            {
                dic.Add(nums[i], i);
            }
        }
        return new int[0];
        // return test (nums, target);
    }

    private int[] test(int[] nums, int target)
    {
        Dictionary<int, int> dic = new Dictionary<int, int>();
        for (int i = 0; i < nums.Length; i++)
        {
            if (dic.ContainsKey(target - nums[i]) && dic[target - nums[i]] != i)
            {
                return new int[2] { dic[target - nums[i]], i };
            }
            if (dic.ContainsKey(nums[i]) == false)
            {
                dic.Add(nums[i], i);
            }
        }
        return new int[0];
    }
}
// @lc code=end